### PR TITLE
Various updates

### PR DESCRIPTION
**Pull request by egpbos**
- Made code device-agnostic by replacing `.cuda()` or `device=cuda:0` with a function which detects whether a GPU is present and specifies to use CPU otherwise.
- Switched most configuration (paths and other options) to use ConfigArgParse
- Fixed some undefined name bugs, **please review the following**:
    * platalea/decoders.py: added initial value of `-np.inf` for undefined value, does this value make sense?
    * ~platalea/encoders.py: mismatched parameter name, I guess it should be `text`, right?~
- Switched from `os.path` to `pathlib` in two modules
- ~Also **please review** utils/copy_best.py, a line was never reached, was this intended?~

Dev-infra:
- Added CI "testing" (running basic-stack experiment) with GitHub actions
- Includes coverage using codecov.io
- Added .flake8 for linting and generic Python .gitignore

### CODE REVIEW

**Comment by egpbos** [comment]
Ok, reverted the undefined name errors which @bhigy fixed in commits of yesterday.

@bhigy could you review whether the fix here makes sense? https://github.com/gchrupala/platalea/pull/3/files#diff-2865c360a054bf744d79a018974f245fR173

File: .flake8 [hunk 1]
```diff
@@ -0,0 +1,2 @@
+[flake8]
```
Comment (bhigy) [inline .flake8]: Is it common to have this kind of configuration as part of the repository? I feel like this should be left to each of us to decide.

File: .flake8 [hunk 2]
```diff
@@ -0,0 +1,2 @@
...
+max-line-length=120
```
Comment (bhigy) [inline .flake8]: Is it common to have this kind of configuration as part of the repository? I feel like this should be left to each of us to decide.

File: .github/workflows/pythonapp.yml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      run: |
+        pip install sklearn python-Levenshtein coverage
+        coverage run experiments/basic-stack/run.py --epochs=1 --meta=dataset.json --audio_features_fn=mfcc_features.pt
+        coverage xml
+
+    - name: Upload coverage to Codecov  
+      uses: codecov/codecov-action@v1
+      with:
+        token: ${{ secrets.CODECOV_TOKEN }}
+        fail_ci_if_error: true
```
Comment (bhigy) [inline .github/workflows/pythonapp.yml:58]: Can you quickly explain what this is doing?

File: platalea/dataset.py [hunk 4]
```diff
@@ -52,17 +53,19 @@ def __init__(self, root, feature_fname, split='train', language='en',
...
         if downsampling_factor is not None:
             num_examples = len(self.metadata) // downsampling_factor
             self.metadata = random.sample(self.metadata, num_examples)
         # mapping from image id to list of caption id
         self.image_captions = {}
-        for line in open(os.path.join(root + 'flickr_audio/wav2capt.txt')):
-            audio_id, image_id, text_id = line.split()
-            text_id = int(text_id[1:])
-            self.image_captions[image_id] = self.image_captions.get(image_id, []) + [(text_id, audio_id)]
+        with open(root_path / 'wav2capt.txt') as fwav2capt:
```
Comment (bhigy) [inline platalea/dataset.py]: wav2capt.txt is by default under flickr_audio folder
```suggestion
        with open(root_path / 'flickr_audio' / 'wav2capt.txt') as fwav2capt:
```

File: platalea/preprocessing.py [hunk 5]
```diff
@@ -7,40 +6,41 @@
...
+    flickr8k_image_features(images_config)
 
 
 def flickr8k_audio_features(config):
-    directory = os.path.join(config['dir'], 'flickr_audio/wavs')
-    wav2capt_fname = os.path.join(config['dir'], 'flickr_audio/wav2capt.txt')
-    files = [line.split()[0] for line in open(wav2capt_fname)]
-    paths = [os.path.join(directory, file) for file in files]
+    directory = config['dataset_path'] / config['audio_subdir']
+    files = [line.split()[0] for line in open(config['dataset_path'] / 'wav2capt.txt')]
```
Comment (bhigy) [inline platalea/preprocessing.py]: Same as before, wav2capt.txt is by default under flickr_audio folder
```suggestion
    files = [line.split()[0] for line in open(config['dataset_path'] / 'flickr_audio' / 'wav2capt.txt')]
```

File: platalea/decoders.py [hunk 6]
```diff
@@ -170,6 +170,7 @@ def beam_search(self, encoder_outputs, beam_size):
                 # Filter ended sequences
                 ended = (new_hyps[:, -1] == self.eos_id).nonzero()[0]
                 best_ended_idx = None
+                best_ended_score = -np.inf  # TODO: please review whether this initial value makes sense (need something, otherwise undefined in if statement below, except that best_ended_hyp is always set to none on line 134 above, but I guess that may change? If not, then the if statement is perhaps not useful and could be removed?)
```
Comment (bhigy) [inline platalea/decoders.py]: The check `best_ended_hyp is None` prevents evaluation of `new_scores[ended[0]] > best_ended_score` before `best_ended_score` is set but setting it to -inf is probably safest
```suggestion
                best_ended_score = -np.inf
```

File: .gitignore [hunk 7]
```diff
@@ -3,4 +3,144 @@
...
+# mypy
+.mypy_cache/
+.dmypy.json
+dmypy.json
+
+# Pyre type checker
+.pyre/
+
+# pytype static type analyzer
+.pytype/
```
Comment (bhigy) [inline .gitignore:146]: Is everything useful here? I feel like there are many things we will never use. Not that it is really problematic to have it but I usually try to keep things to the minimum.

File: README.md [hunk 8]
```diff
@@ -1,2 +1,50 @@
...
+
+
+## Install
+
+```sh
+conda create -n platalea python==3.8 pytorch -c conda-forge -c pytorch
+conda activate platalea
+pip install torchvision
+pip install .
+```
```
Comment (bhigy) [inline README.md:16]: Shouldn't we use setup.py in the install process? see e.g. https://github.com/gchrupala/symbolic-bias

File: README.md [hunk 9]
```diff
@@ -1,2 +1,50 @@
...
+### Configuration
+We use ConfigArgParse for setting necessary input variables.
+This means you can use either a configuration file (config.ini or config.yml), environment variables or command line arguments to specify the necessary configuration parameters.
+For instance, to use environment variables with the Flickr1d dataset, set the PLATALEA_DATA_ROOT variable to the directory in which the Flickr1d data was downloaded, e.g. if it was cloned at `/user/data/flickr1d` do:
+
+```sh
+export PLATALEA_DATA_ROOT=/user/data/flickr1d
+export PLATALEA_METADATA_JSON=dataset.json
+export PLATALEA_AUDIO_FEATURES_FN=mfcc_features.pt
+```
```
Comment (bhigy) [inline README.md:33]: How is CI set up? Through command line parameters? Would a flickr1d.yml config file make sense for switching easily between the datasets?

File: .github/workflows/pythonapp.yml (hunk 3 above)
Comment (egpbos) [reply .github/workflows/pythonapp.yml:58]: This is from here: https://github.com/codecov/codecov-action It uploads the coverage report that is generated in the previous action where I replaced `python` with `coverage run`. `coverage run` is just a wrapper around `python` that runs the code, but also keeps track of what parts of the code are used.

File: .flake8 (hunk 2 above)
Comment (egpbos) [reply .flake8]: You're right, I'll revert that.

File: platalea/dataset.py (hunk 4 above)
Comment (egpbos) [reply platalea/dataset.py]: Ah, yes, actually it was both in the flickr8k folder and in the flickr_audio sub folder, they were duplicates. I'm fine with either choice, under audio makes sense. Should then change that in flickr1d as well.

File: platalea/preprocessing.py (hunk 5 above)
Comment (egpbos) [reply platalea/preprocessing.py]: Same as before indeed.

File: platalea/decoders.py (hunk 6 above)
Comment (egpbos) [reply platalea/decoders.py]: Yes indeed, it's never reached in the current setup, that's why Python never triggered an exception on this, but the linter did pick it up, apparently :)

File: .gitignore (hunk 7 above)
Comment (egpbos) [reply .gitignore:146]: Surely not everything is useful in our project right now. It's just easy to copy paste generic catch-all Python project gitignores, otherwise, you tend to run into things later on. For instance, we may want to use some notebooks at some point, or maybe we'll need some Cython at some point (probably not, but who knows).

File: README.md (hunk 8 above)
Comment (egpbos) [reply README.md:16]: This is equivalent, `pip install .` also uses the information in setup.py. The advantage is that it's also properly registered in pip, which may be important for dependencies. And it's easier to uninstall, since setup.py uses `easy_install` for installing, which often doesn't clean up well, whereas with pip you don't have any such problems. Don't think it matters much in our case, but in general, nowadays `pip install .` seems to be preferred over `python setup.py install`.

File: README.md (hunk 9 above)
Comment (egpbos) [reply README.md:33]: Great idea! Should add that to the flickr1d repo. I'll add an issue there to keep track of the todo's, also the change of directories for the wav2...txt files.

File: .flake8 (hunk 2 above)
Comment (egpbos) [reply .flake8]: done.

File: README.md (hunk 8 above)
Comment (bhigy) [reply README.md:16]: Great!

**Review by bhigy — APPROVED** [review]

File: platalea/basic.py [hunk 10]
```diff
@@ -85,7 +90,7 @@ def val_loss():
         for epoch in range(1, config['epochs']+1):
             cost = Counter()
             for j, item in enumerate(data['train'], start=1): # check reshuffling
-                item = {key: value.cuda() for key, value in item.items()}
+                item = {key: value.to(_device) for key, value in item.items()}
                 loss = net.cost(item)
                 optimizer.zero_grad()
                 loss.backward()
```
Comment (bhigy) [inline platalea/basic.py:96]: Using device should be done everywhere `cuda()` is used to be consistent.

**Review by bhigy — CHANGES_REQUESTED** [review]

File: platalea/basic.py (hunk 10 above)
Comment (egpbos) [reply platalea/basic.py:96]: Ah sure, weird that i didn't catch that one before.
